solana-programs/anchor/audius-data/programs/audius-data/src/lib.rs

```diff
@@ -95,10 +99,250 @@ pub mod audius_data {
+        let proposers = [
+            (&ctx.accounts.proposer1, &ctx.accounts.proposer1_authority),
+            (&ctx.accounts.proposer2, &ctx.accounts.proposer2_authority),
+            (&ctx.accounts.proposer3, &ctx.accounts.proposer3_authority),
+        ];
+
+        // Ensure that no proposer's owner eth address is repeated
+        let mut eth_addresses = BTreeMap::new();
+        if !proposers.iter().all(move |(proposer, authority)| return match eth_addresses.insert(proposer.owner_eth_address, true) {
+            Some(_) => false,
+            None => proposer.authority == authority.key()
+        }) {
+            // duplicate owner eth address - err
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        // Confirm that the base used for user account seed is derived from this Audius admin storage account
+        let (derived_base, _) = Pubkey::find_program_address(
+            &[&ctx.accounts.admin.key().to_bytes()[..32]],
+            ctx.program_id,
+        );
+
+        if derived_base != base {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        // Refer to context here - https://docs.solana.com/developing/programming-model/transactions#multiple-instructions-in-a-single-transaction
+        let dummy_owner_field = Pubkey::from_str("11111111111111111111111111111111").unwrap();
+
+        let content_node = &mut ctx.accounts.content_node;
+        content_node.authority = dummy_owner_field;
+
+        Ok(())
+    }
+
+    /// Update a user's replica set
+    pub fn update_user_replica_set(
+        ctx: Context<UpdateUserReplicaSet>,
+        base: Pubkey,
+        _user_handle: UserHandle,
+        replica_set: [u16;3],
+        _replica_set_bumps: [u8; 3]
+    ) -> Result<()> {
+
+        // Confirm that the base used for user account seed is derived from this Audius admin storage account
+        let (derived_base, _) = Pubkey::find_program_address(
+            &[&ctx.accounts.admin.key().to_bytes()[..32]],
+            ctx.program_id,
+        );
+
+        if derived_base != base {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        let user_replica_set = &ctx.accounts.user.replica_set;
+
+        let did_user_sign = ctx.accounts.user.authority == ctx.accounts.cn_authority.key();
+        let mut is_valid_authority = did_user_sign;
+        let proposers = [&ctx.accounts.cn1, &ctx.accounts.cn2, &ctx.accounts.cn3];
+
+        // Validate that a content node in the replica signed the tx if the user's authority did not
+        if !is_valid_authority {
+            for proposer in proposers {
+                if proposer.authority == ctx.accounts.cn_authority.key() {
+                    is_valid_authority = true;
+                    // Validate that one of the user's replica set nodes signed the transaction
+                    if !user_replica_set.iter().any(|cn| {
+                        let (base_pda, _bump) = Pubkey::find_program_address(
```

Comment:
i know this is a NIT but i think base_pda here is very misleading - this is the cn_account_pda right?

Reply:
also we could validate the bump too right?

Reply:
I don't think we need to verify the bump here because we are only deriving the base pda from the user's replica set - so these should have already been verified. it's used to compare the pda against the input cn account's pda. 
Let me know if this make sense and we can talk offline